Web apps datetimepicker fixes

orangejenny opened the pull request:
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-2248

## Safety Assurance

### Safety story
UI-level changes only, although web apps js and this widget in particular are finicky. I'll ask QA to bang on this change and see if they can break anything.

### Automated test coverage

Tests for the date calculation are in this PR. No UI tests for the widget.

### QA Plan

Requesting QA.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Improved blurring behavior

orangejenny committed:
Added new logic for converting two-digit years

orangejenny committed:
Fixed lint error

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/entries_spec.js:
```diff
@@ -325,6 +326,13 @@ describe('Entries', function () {
         assert.isTrue(spy.calledOnce);
     });
 
+    it('Should convert two-digit dates to four-digit dates', function () {
```
:trophy:

orangejenny committed:
Fixed lint errors and removed guard

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/entries_spec.js:
```diff
@@ -34,7 +35,7 @@ describe('Entries', function () {
         spy = sinon.spy();
         $.subscribe('formplayer.' + Const.ANSWER, spy);
-        this.clock = sinon.useFakeTimers();
+        this.clock = sinon.useFakeTimers(new Date("2020-03-15 15:41").getTime());
```
this would've been a fun test time bomb to start triggering in 6 years

orangejenny replied:
💣

orangejenny commented on corehq/apps/cloudcare/static/cloudcare/js/util.js:
```diff
@@ -282,6 +282,31 @@ hqDefine('cloudcare/js/util', [
+            if (inputYear.length === 2) {
+                inputYear = Math.floor(new Date().getFullYear() / 100) + inputYear;
+                if (inputYear > new Date().getFullYear() + 10) {
+                    inputYear -= 100;
```
The logic here is that line 300 assumes that the given date is in the current century, and then the if statement checks to see if that calculated date is "too far" into the future (10+ years), and if so, moves it back to the previous century. Open to feedback on making it clearer.

orangejenny committed:
Added guard against nonexistent el

orangejenny committed:
Fixed incomplete rename

orangejenny committed:
Fixed logic for parsing date once two-digit year has been converted.

I'm not sure if or how the previous code ever worked: parseInputDate doesn't
give access to the default implementation of parseINputDate, only to the
option the user passed in (if any). The only way I found to access the default
implementation is to call the "private" function, which is questionable, but
better than re-implementing the widget's default parsing.

orangejenny committed:
Added handling of invalid dates

orangejenny committed:
Fixed parseInputDate override getting ignored
